### PR TITLE
[MPOM-264] Parameterize maven-compiler-plugin with parameter "release"

### PR DESCRIPTION
**Pull request by kwin**
when running on JDK 9+

### CODE REVIEW

**Comment by rfscholte** [comment]
I think this is too dangerous. What if projects are using toolchains?

File: pom.xml [hunk 1]
```diff
@@ -370,6 +370,44 @@ under the License.
...
+            <executions>
+              <execution>
+                <id>set-compiler-release-parameter</id>
+                <goals>
+                  <goal>regex-property</goal>
+                </goals>
+                <configuration>
+                  <name>maven.compiler.release</name>
+                  <!-- strip leading "1." from maven.compiler.target -->
+                  <regex>1\.(.*)</regex>
```
Comment (michael-o) [inline pom.xml:400]: What if the target contains `11`? This is too much magic here...

Comment (kwin) [reply pom.xml:400]: If the value doesn't start with "1." it is just used as is.

Comment (kwin) [reply pom.xml:400]: I am open for other proposals on how to achieve this while keeping backwards compatibility and only being forced to define the target java version in one property.

Comment (michael-o) [reply pom.xml:400]: Thanks for the information.

**Comment by kwin** [comment]
> I think this is too dangerous. What if projects are using toolchains?

Is it possible to add a profile condition for toolchains?

**Comment by rfscholte** [comment]
No, the JDK activation only works for the Maven runtime.

**Comment by kwin** [comment]
@rfscholte Any proposal how to put this functionality into ASF parent pom or do you rather think every project should configure "release" on a project level.

**Comment by rfscholte** [comment]
This is a returning discussion: if we need to update source/target(/release) it means there is just no sane default value. IMO projects should always define these values themselves to prevent surprises by updating a parent.

**Comment by kwin** [comment]
This PR is not about defining the default value but about setting the target java version in one(!) property and leverage it both in Java < 9 and 9 + optimally

**Comment by rfscholte** [comment]
Toolchain will still be an issue.

**Comment by rfscholte** [comment]
Another argument: if both Michael and I don't see it right away what's happening here, it is "magic" that should happen inside a plugin (where the magic can be tested). It will make the pom too hard to maintain (which would bring us back to the Ant buildscript-master: only one person will understand what is happening here, there rest either ignores it or doesn't dare to touch it)

**Comment by kwin** [comment]
I opened https://github.com/codehaus-plexus/plexus-compiler/issues/140 to allow to set the parameter "maven.compiler.release" without risking to break the build for JDK < 9.

**Comment by jira-importer** [comment]
Resolve #479

**Pull request closed without merging**